packages/block-editor/src/components/provider/index.js

```diff
@@ -1,139 +1,88 @@
+import { useSelect, useDispatch } from '@wordpress/data';
 
 /**
  * Internal dependencies
  */
 import withRegistryProvider from './with-registry-provider';
 
-class BlockEditorProvider extends Component {
-	componentDidMount() {
-		this.props.updateSettings( this.props.settings );
-		this.props.resetBlocks( this.props.value );
-		this.attachChangeObserver( this.props.registry );
-	}
+function BlockEditorProvider( {
+	settings,
+	value,
+	children,
+	onChange = () => {},
+	onInput = () => {},
+} ) {
+	const { updateSettings, resetBlocks } = useDispatch( 'core/block-editor' );
 
-	componentDidUpdate( prevProps ) {
-		const {
-			settings,
-			updateSettings,
-			value,
-			resetBlocks,
-			registry,
-		} = this.props;
-
-		if ( settings !== prevProps.settings ) {
-			updateSettings( settings );
-		}
-
-		if ( registry !== prevProps.registry ) {
-			this.attachChangeObserver( registry );
-		}
-
-		if ( this.isSyncingOutcomingValue ) {
-			this.isSyncingOutcomingValue = false;
-		} else if ( value !== prevProps.value ) {
-			this.isSyncingIncomingValue = true;
-			resetBlocks( value );
-		}
-	}
-
-	componentWillUnmount() {
-		if ( this.unsubscribe ) {
-			this.unsubscribe();
-		}
-	}
-
-	/**
-	 * Given a registry object, overrides the default dispatch behavior for the
-	 * `core/block-editor` store to interpret a state change and decide whether
-	 * we should call `onChange` or `onInput` depending on whether the change
-	 * is persistent or not.
-	 *
-	 * This needs to be done synchronously after state changes (instead of using
-	 * `componentDidUpdate`) in order to avoid batching these changes.
-	 *
-	 * @param {WPDataRegistry} registry     Registry from which block editor
-	 *                                      dispatch is to be overriden.
-	 */
-	attachChangeObserver( registry ) {
-		if ( this.unsubscribe ) {
-			this.unsubscribe();
-		}
+	useEffect( () => {
+		updateSettings( settings );
+	}, [ settings ] );
 
+	const { blocks, isPersistent, isIgnored } = useSelect( ( select ) => {
 		const {
 			getBlocks,
 			isLastBlockChangePersistent,
 			__unstableIsLastBlockChangeIgnored,
-		} = registry.select( 'core/block-editor' );
+		} = select( 'core/block-editor' );
 
-		let blocks = getBlocks();
-		let isPersistent = isLastBlockChangePersistent();
-
-		this.unsubscribe = registry.subscribe( () => {
-			const {
-				onChange,
-				onInput,
-			} = this.props;
-			const newBlocks = getBlocks();
-			const newIsPersistent = isLastBlockChangePersistent();
-			if (
-				newBlocks !== blocks && (
-					this.isSyncingIncomingValue ||
-					__unstableIsLastBlockChangeIgnored()
-				)
-			) {
-				this.isSyncingIncomingValue = false;
-				blocks = newBlocks;
-				isPersistent = newIsPersistent;
-				return;
+		return {
+			blocks: getBlocks(),
+			isPersistent: isLastBlockChangePersistent(),
+			isIgnored: __unstableIsLastBlockChangeIgnored(),
+		};
+	} );
+
+	const previousBlocks = useRef( blocks );
```

Comment:
This looks like a use-case for a [`usePrevious`](https://usehooks.com/usePrevious/) hook. I'm actually using one already in #21181, and I wonder if we should add it to `@wordpress/compose`.

Reply:
`usePrevious` has now been added to the `@wordpress/compose` package, so it could definitely be used here.